src/net.cpp

```diff
@@ -918,14 +919,23 @@ bool CConnman::AttemptToEvictConnection()
     if (!node_id_to_evict) {
         return false;
     }
-    LOCK(m_nodes_mutex);
-    for (CNode* pnode : m_nodes) {
-        if (pnode->GetId() == *node_id_to_evict) {
-            LogPrint(BCLog::NET, "selected %s connection for eviction peer=%d; disconnecting\n", pnode->ConnectionTypeAsString(), pnode->GetId());
-            pnode->fDisconnect = true;
-            return true;
+    CNode* evicted_node{nullptr};
+    {
+        LOCK(m_nodes_mutex);
+        for (CNode* pnode : m_nodes) {
+            if (pnode->GetId() == *node_id_to_evict) {
+                LogPrint(BCLog::NET, "selected %s connection for eviction peer=%d; disconnecting\n", pnode->ConnectionTypeAsString(), pnode->GetId());
+                pnode->fDisconnect = true;
+                evicted_node = pnode;
+                break;
+            }
         }
     }
+    if (evicted_node) {
+        DisconnectAndReleaseNode(evicted_node);
+        DeleteDisconnectedNode(evicted_node);
```

Comment:
`DisconnectAndReleaseNode()` is safe because its body is protected by `m_nodes_mutex`. But `DeleteDisconnectedNode()` is not - two threads could race to call `delete` on the `CNode` object. It looks ok to iterate the whole `m_disconnected_nodes` here to avoid that, like in the other place.